binary_search_tree/tree.py

```diff
@@ -7,53 +10,129 @@ def __init__(self, key, val = None):
         self.value = val
         self.left = None
         self.right = None
-        
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    # Time Complexity: O(log n)
+    # Space Complexity: O(n)
+    def add(self, key, value = None):        
+        if not self.root:
+            self.root = TreeNode(key, value)
+            return
+        new_node = TreeNode(key, value)
+        current = self.root
+        while current: 
+            if new_node.key < current.key:
+                if current.left == None:
+                    current.left = new_node
+                    return
+                else:
+                    current = current.left 
+            else: 
+                if current.right == None:
+                    current.right = new_node
+                    return
+                else:
+                    current = current.right
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        current = self.root
+        while current:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+        return None     
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        values = []
+        self.inorder_helper(self.root, values)
+        return values
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)   
     def preorder(self):
-        pass
+        values = []
+        self.preorder_helper(self.root, values)
+        return values
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def postorder(self):
-        pass
+        values = []
+        self.postorder_helper(self.root, values)
+        return values
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def height(self):
```

Comment:
✨ Yes! Time and space complexity would be O(n) if the tree is unbalanced. If the tree is balanced, time complexity would be O(log n)